types/src/transaction/mod.rs

```diff
@@ -837,7 +835,15 @@ impl From<VMStatus> for TransactionStatus {
                 },
                 _ => TransactionStatus::Keep(recorded.into()),
             },
-            Err(code) => TransactionStatus::Discard(code),
+            Err(code) => {
+                if code.status_type() == StatusType::InvariantViolation
+                    && charge_invariant_violation
+                {
+                    TransactionStatus::Keep(ExecutionStatus::MiscellaneousError(Some(code)))
```

Comment:
It might be cleaner to add a separate variant `ExecutionStatus::InvariantViolation` since `ExecutionStatus::MiscellaneousError` has been historically used for execution errors that can be triggered by user inputs. Don't know if we'll run into issues with this though. The whole error conversion story is a mess that was left out in the previous rounds of refactoring.

Reply:
I'm a bit concerned about adding another category cuz the storage might have trouble deserializing this new error category

Reply:
Why would storage have trouble deserializing this if we change all the locations that look at the different variations of ExecutionStatus?